canal.cc

```diff
@@ -313,49 +358,49 @@ mainExcept(int argc, char *argv[])
                 }
             }
         } else {
-            for (auto loc = hdr.p_vaddr; loc < hdr.p_vaddr + hdr.p_filesz; loc += sizeof p) {
-                // log a '.' every megabyte.
-                if (verbose && (loc - hdr.p_vaddr) % (1024 * 1024) == 0)
+            const size_t step = sizeof(Elf::Off);
+            const size_t chunk_size = 1'048'576;
+            Elf::Addr loc=hdr.p_vaddr;
+            const Elf::Addr end_loc = loc + hdr.p_filesz;
+            while (loc < end_loc) {
+                size_t read_size = std::min(chunk_size, end_loc - loc);
+                data.resize(read_size/step);
+                read_size = process->io->read(loc, read_size, reinterpret_cast<char*>(data.data()));
+                data.resize(read_size / step);
+                if (verbose) {
+                    // log a '.' every megabyte.
                     clog << '.';
-                try {
-                    process->io->readObj(loc, &p);
                 }
-                catch (...) {
-                    *debug << "read failed\n";
-                    break;
-		}
-                if (searchaddrs.size()) {
-                    for (auto range = searchaddrs.begin(); range != searchaddrs.end(); ++range) {
-                        if (p >= range->first && p < range->second ) {
-			    searchmatches++;
-			    if (verbose) {
-				IOFlagSave _(cout);
-				cout << "0x" << hex << loc << "\n";
-			    }
+                for (auto it=data.begin(); it != data.end(); ++it, loc+=step) {
+                    const auto & p=*it;
+                    if (searchaddrs.size()) {
+                        for (auto range = searchaddrs.begin(); range != searchaddrs.end(); ++range) {
+                            if (p >= range->first && p < range->second && (p % 4 == 0)) {
+                                IOFlagSave _(cout);
+                                cout << "0x" << hex << loc << "\n";
+                            }
                         }
-                    }
-                } else {
-                    auto found = lower_bound(listed.begin(), listed.end(), p);
-                    if (found != listed.end() &&
-                            (symOffset != -1
-                                ? found->memaddr() + symOffset == p
-                                : found->memaddr() <= p && found->memaddr() +
-                                               found->sym.st_size > p)) {
-                        if (showaddrs)
-                            cout
-                                << found->name << " 0x" << std::hex << loc
-                                << std::dec <<  " ... size=" << found->sym.st_size
-                                << ", diff=" << p - found->memaddr() << endl;
+                    } else {
+                        auto tuple = store.find(p);
+                        auto found = std::get<0>(tuple);
```

Comment:
you could use std::tie to make this clearer, but if you want to, you can follow up this PR